ezmlm-confirm.1
.TH ezmlm-confirm 1
.SH NAME
ezmlm-confirm \- process user actions for message
confirm and discard
.SH SYNOPSIS
.B ezmlm-confirm [-cCrRvV]
.I dir ['/path/program args']

.SH DESCRIPTION
.B ezmlm-confirm
reads a mail message, expecting it to contain an
.I -confirm
or 
.I -discard
action request for the mailing list stored in
.I dir.

.B ezmlm-confirm
verifies the cookie sent as part of the action request, and if correct
attempts to find the corresponding message in
.IR dir\fB/mod/unconfirmed/ .
If the message is found, it is either discarded or posted to the list,
depending on the action request.

Posts to the list are handled by
piping the message to
.B ezmlm-send(1)
located in the ezmlm binary directory, as set at compile time. This is
usually the directory that ezmlm-moderate resides in.
.B ezmlm-send(1)
is provided with
.I dir
as the first argument.

If the optional second argument is given,
.B ezmlm-confirm
pipes the message to that program, as executed by sh.
No parameters are supplied.

.I dir
is passed as an argument to
.BR ezmlm-send(1) ,
but NOT to any program specified as the optional
.B ezmlm-confirm
command line argument.

.B ezmlm-confirm
does not bother to correctly set sender.
.B ezmlm-send(1)
doesn't care, and any other program that might be used can parse the
sender from the first line of the message, which is always
.I Return-Path: <address@host.domain>
as build from SENDER originally passed to
.BR ezmlm-store(1) .

Once the message has been successfully accepted or discarded, it is removed from
.IR dir\fB/mod/unconfirmed/ .

If the message is not found in
.IR dir\fB/mod/unconfirmed/ ,
it has timed out and the message or
the message stub has been removed by
.BR ezmlm-clean(1) .
In this case,
.B ezmlm-confirm
notifies the sender of the failure.
Notification is in the form of a fatal error from qmail.

At the beginning of the message,
.B ezmlm-confirm
prints a new
.B Mailing-List
field with the contents of the
.I TXT_MAILING_LIST
message.  It rejects the message if there is already a
.B Mailing-List
field.

.B ezmlm-confirm
does not distribute bounce messages:
if the environment variable
.B SENDER
is set, and is either empty or
.BR #@[] ,
.B ezmlm-confirm
rejects the message.
.SH OPTIONS
.TP
.B \-cCrR
Passed on to ezmlm-send(1).
.TP
.B \-v
Display
.B ezmlm-confirm
version information.
.TP
.B \-V
Display
.B ezmlm-confirm
version information.
.SH "CHARACTER SETS"
If
.I dir\fB/charset
exists,
.B ezmlm-confirm
will use the character set listed for all messages. Otherwise, the
default ``us-ascii'' will be used. The character set can be suffixed
by ``:'' followed by a code. If the code is ``Q'', outgoing messages are 
sent as ``Quoted-Printable'', if it is ``B'' they are sent ``base64'' encoded.
Otherwise, text is sent as is.
.SH "SEE ALSO"
ezmlm-moderate(1),
ezmlm-clean(1),
ezmlm-make(1),
ezmlm-send(1),
ezmlm-store(1),
ezmlm(5)


CHANGES
BUGS:	-ezmlm-idx does not implement a full rfc822 parser. Comments can be
	 inserted to defeat the MIME parsing mechanism, such as:
	 Content-Type: (comment) test/plain
	 This type of comment, while legal, does not occur in normal messages.
	 Guiding info past ezmlm-idx MIME parsing is easier done by enclosing
	 a message within a message. The mechanism isn't intended as a
	 "secure" way of preventing e.g. "application/ms-tnef" from being
	 disseminated.
	 If you think this is a problem, E-mail ezmlm@list.cr.yp.to.
	 Mimeremove/reject work only at the outer MIME layer.
	-ezmlm-request requires addresses to be user@host. While rfc822
	 compliant, addresses like ``"the man" @ host'' are not accepted.
	-ezmlm-cgi doesn't deal optimally with holes in the archive.
	-ezmlm-cgi seems not to work with Chinese. The problem has been to
	 find someone willing/able to participate in debugging.

Some ezmlmrc are not updated and may not work right.  If you fix them,
please send them to ezmlm@list.cr.yp.to for inclusion in the next
release.

ezmlm-idx-7.1.0, ????-??-??
===========================

- Added support for wildcard addresses in all address databases.

- Added support for qmqpservers to all ezmlm-idx programs.

- Modified the (un)subscribe procedure to not require confirmation
  before telling the sender they were already on (or off) the list.

ezmlm-idx-7.0.2, 2009-08-09
===========================

- Fixed selection of [un]subscribe subject line for moderated
  subscriptions.

- Clarified ezmlm-reject man page to note that Precedence: bulk is also
  rejected.

- Added a proper charset for ch_GB (Chinese GuoBiao), and added an
  explicit charset for the other languages that were missing them.

- Switched all ISO-8859-1 charsets to the more modern ISO-8859-15.

- Many small text updates and tweaks.

ezmlm-idx-7.0.1, 2009-07-09
===========================

- Added support for decoding sender addresses mangled by BATV prvs.

- Added missing ezmlm-checksub to the installed programs.

- Improved the TXT_BY message used in digests to work better with
  non-English languages.

- Fixed out-of-memory resulting in a permanent error.

- Fixed a buffer underflow bug in concatHDR.

- Fixed a bug in ezmlm-issubn.c that was causing it to scan the main
  list when it shouldn't.

- Fixed missing defaults in ezmlm-tstdig when digsize, digcount, or
  digtime were not created.

- Clarified the steps necessary when upgrading in UPGRADE.

ezmlm-idx-7.0.0, 2008-06-16
===========================

This version has three major architectural changes:

1. There is now support for internationalized messages.  All error
messages, all subject lines, and a few other messages are now configured
in a file named "text/messages" that is read at run time.  Messages in
this file will undergo header-style substitution before they are output.

Unlike other text files, all 3 of the files (DIR/text/messages,
/etc/ezmlm/LANG/text/messages, and /etc/ezmlm/default/text/messages) are
read, and only the first match (in the above order) is used for any
given message.  This allows for creation of partial files to override
just select messages.  If no match is found, the internal English text
is used.

2. The use of the "flags" file has been deprecated completely by
individual flag files accessed by the appropriate programs.  This, along
with corresponding changes to the ezmlmrc template, allows for lists to
be reconfigured without invoking ezmlm-make.  To upgrade a list to the
new setup, simply run "ezmlm-make -+ DIR".  Note that this will delete
the "config" and "flags" files, as they have been superceded by other
files.

3. Creation of subscriber tables has been moved out of ezmlm-mktab-* and
into ezmlm-make by adding some additional hooks to the subdb plugins.  A
ezmlm-rmtab program is also added to remove subscriber tables, also with
hooks in the subdb plugins.  The ezmlm-mktab-* programs have been
obsoleted by these changes and have been removed.

Smaller changes:

- Fixed a bug in ezmlm-request that caused a segfault when attempting to
  execute the "which" command.

- Added new program ezmlm-checksub to replace the
	ezmlm-issubn ... || { echo error; exit 100; }
  lines in the ezmlmrc files.  ezmlm-issubn is now deprecated for use in
  .qmail files, but still exists for backwards compatibility.

- Added ezmlm-weed to the confirmer control files.

- Added <#a#> substitution for the local part of the accept address.

- ezmlm-warn will now process bounces for both the main and digest lists
  if neither the -d nor the -D options are given.

- Added several control files for programs:
	dir/digcount		ezmlm-tstdig
	dir/digestcode		ezmlm-get
	dir/digformat		ezmlm-get
	dir/digsize		ezmlm-tstdig
	dir/digtime		ezmlm-tstdig
	dir/modcanedit		ezmlm-manage
	dir/modcanlist		ezmlm-manage
	dir/modgetonly		ezmlm-get
	dir/modpostonly		ezmlm-store
	dir/nosubconfirm	ezmlm-manage
	dir/nounsubconfirm	ezmlm-manage
	dir/nowarn		ezmlm-warn
	dir/subgetonly		ezmlm-get
	dir/noreturnposts	ezmlm-clean

- Removed the ezmlm-make -4 option in favor of the above files.

- Fixed handling of digest bouncer in ezmlm-dispatch.

- Eliminated the use of vfork for portability.

- Added support for decoding sender addresses mangled by SRS.

- Fixed bug when running ezmlm-archive on a newly-created list.

- Always enable ezmlm-request in manager.

- Always enable the "deny" blacklist.

- Fixed a long standing bug in ezmlm-store that caused both posting and
  moderating to happen when both modpost and confirmpost were disabled.

Note: The subdb API was modified in this version.  You will need to
reinstall all sub-* modules along with the main package.  Also, the
messages changes also replaced the confirmpost subject line hack in
ezmlm-store and the mailinglist file.  The contents of DIR/confirmpost
and DIR/mailinglist are now ignored in favor of the SUB_CONFIRM_POST and
TXT_MAILING_LIST messages, respectively.

ezmlm-idx-6.0.1, 2007-10-06
===========================


ezmlm-confirm.c
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include "error.h"
#include "case.h"
#include "stralloc.h"
#include "str.h"
#include "env.h"
#include "error.h"
#include "sig.h"
#include "fork.h"
#include "wait.h"
#include "strerr.h"
#include "byte.h"
#include "getln.h"
#include "qmail.h"
#include "substdio.h"
#include "readwrite.h"
#include "seek.h"
#include "quote.h"
#include "datetime.h"
#include "now.h"
#include "fmt.h"
#include "sgetopt.h"
#include "cookie.h"
#include "messages.h"
#include "copy.h"
#include "open.h"
#include "lock.h"
#include "wrap.h"
#include "die.h"
#include "idx.h"
#include "mime.h"
#include "config.h"
#include "auto_version.h"

static int flagmime = MOD_MIME;	/* default is message as attachment */

const char FATAL[] = "ezmlm-confirm: fatal: ";
const char INFO[] = "ezmlm-confirm: info: ";
const char USAGE[] =
"ezmlm-confirm: usage: ezmlm-confirm [-cCmMrRvV] dir [/path/ezmlm-send]";

static stralloc to = {0};
static stralloc sendopt = {0};
static datetime_sec when;

static char hash[COOKIE];
static stralloc line = {0};
static stralloc fnbase = {0};
static stralloc fnmsg = {0};

static const char *dir;

static struct stat st;

static int checkfile(const char *fn)
/* looks for DIR/mod/unconfirmed/fn.              */
/* Returns:                                       */
/*          1 found in unconfirmed                */
/*          0 not found                           */
/* Handles errors.                                */
/* ALSO: if found, fnmsg contains the o-terminated*/
/* file name.                                     */
{
  if (!stralloc_copys(&fnmsg,"mod/unconfirmed/")) die_nomem();
  if (!stralloc_cats(&fnmsg,fn)) die_nomem();
  if (!stralloc_0(&fnmsg)) die_nomem();
  if (stat(fnmsg.s,&st) == 0)
    return 1;
  if (errno != error_noent)
    strerr_die2sys(111,FATAL,MSG1(ERR_STAT,fnmsg.s));
  return 0;
}

static substdio sstext;
static char textbuf[1024];

static void maketo(void)
/* expects line to be a return-path line. If it is and the format is valid */
/* to is set to to the sender. Otherwise, to is left untouched. Assuming   */
/* to is empty to start with, it will remain empty if no sender is found.  */
{
  unsigned int x, y;

  if (case_startb(line.s,line.len,"return-path:")) {
    x = 12 + byte_chr(line.s + 12,line.len-12,'<');
    if (x != line.len) {
      y = byte_rchr(line.s + x,line.len-x,'>');
      if (y + x != line.len) {
        if (!stralloc_copyb(&to,line.s+x+1,y-1)) die_nomem();
        if (!stralloc_0(&to)) die_nomem();
      }			/* no return path-> no addressee. A NUL in the sender */
    }			/* is no worse than a faked sender, so no problem */
  }
}

void main(int argc, char **argv)
{
  const char *sender;
  const char *def;
  const char *local;
  const char *action;
  int fd;
  int match;
  char szchar[2] = "-";
  const char *replyto = (char *) 0;
  unsigned int start,confnum;
  int child;
  int opt;
  const char *cp;

  (void) umask(022);
  sig_pipeignore();
  when = now();

  if (!stralloc_copys(&sendopt,"-")) die_nomem();
  while ((opt = getopt(argc,argv,"cCmMrRt:T:vV")) != opteof)
    switch(opt) {	/* pass on ezmlm-send options */
      case 'c':			/* ezmlm-send flags */
      case 'C':
      case 'r':
      case 'R':
        szchar[0] = (char) opt & 0xff;
        if (!stralloc_append(&sendopt,szchar)) die_nomem();
        break;
      case 'm': flagmime = 1; break;
      case 'M': flagmime = 0; break;
      case 't':
      case 'T': if (optarg) replyto = optarg; break;
      case 'v':
      case 'V': strerr_die2x(0,"ezmlm-moderate version: ",auto_version);
      default:
	die_usage();
    }

  startup(dir = argv[optind++]);

  sender = env_get("SENDER");
  if (!sender) strerr_die2x(100,FATAL,MSG(ERR_NOSENDER));
  local = env_get("LOCAL");
  if (!local) strerr_die2x(100,FATAL,MSG(ERR_NOLOCAL));
  def = env_get("DEFAULT");
  if (!def) strerr_die2x(100,FATAL,MSG(ERR_NODEFAULT));

  if (!*sender)
    strerr_die2x(100,FATAL,MSG(ERR_BOUNCE));
  if (!sender[str_chr(sender,'@')])
    strerr_die2x(100,FATAL,MSG(ERR_ANONYMOUS));
  if (str_equal(sender,"#@[]"))
    strerr_die2x(100,FATAL,MSG(ERR_BOUNCE));

  /* local should be >= def, but who knows ... */
  cp = local + str_len(local) - str_len(def) - 2;
  if (cp < local) die_badformat();
  action = local + byte_rchr(local,cp - local,'-');
  if (action == cp) die_badformat();
  action++;

  if (!action[0]) die_badformat();
  if (!str_start(action,ACTION_CONFIRM) && !str_start(action,ACTION_DISCARD))
    die_badformat();
  start = str_chr(action,'-');
  if (!action[start]) die_badformat();
  confnum = 1 + start + str_chr(action + start + 1,'.');
  if (!action[confnum]) die_badformat();
  confnum += 1 + str_chr(action + confnum + 1,'.');
  if (!action[confnum]) die_badformat();
  if (!stralloc_copyb(&fnbase,action+start+1,confnum-start-1)) die_nomem();
  if (!stralloc_0(&fnbase)) die_nomem();
  cookie(hash,key.s,key.len,fnbase.s,"","a");
  if (byte_diff(hash,COOKIE,action+confnum+1))
    die_badformat();

  lockfile("mod/confirmlock");

  if (!checkfile(fnbase.s)) strerr_die2x(100,FATAL,MSG(ERR_MOD_TIMEOUT));
/* Here, we have an existing filename in fnbase with the complete path */
/* from the current dir in fnmsg. */

  if (str_start(action,ACTION_DISCARD)) {
        unlink(fnmsg.s);
        strerr_die1x(0,"ezmlm-confirm: info: post rejected");
  } else if (str_start(action,ACTION_CONFIRM)) {
    fd = open_read(fnmsg.s);
    if (fd == -1) {
      if (errno !=error_noent)
        strerr_die2sys(111,FATAL,MSG1(ERR_OPEN,fnmsg.s));
      else	/* shouldn't happen since we've got lock */
        strerr_die3x(100,FATAL,fnmsg.s,MSG(ERR_MOD_TIMEOUT));
    }

    substdio_fdbuf(&sstext,read,fd,textbuf,sizeof(textbuf));
				/* read "Return-Path:" line */
    if (getln(&sstext,&line,&match,'\n') == -1 || !match)
      strerr_die2sys(111,FATAL,MSG(ERR_READ_INPUT));
    maketo();			/* extract SENDER to "to" */
    env_put2("SENDER",to.s);	/* set SENDER */
    if (seek_begin(fd) == -1)	/* rewind, since we read an entire buffer */
      strerr_die2sys(111,FATAL,MSG1(ERR_SEEK,fnmsg.s));

    if ((child = wrap_fork()) == 0) {
      close(0);
      dup(fd);	/* make fnmsg.s stdin */
      if (argc > optind + 1)
	wrap_execvp((const char **)argv + optind);
      else if (argc > optind)
        wrap_execsh(argv[optind]);
      else
        wrap_execbin("/ezmlm-send", &sendopt, dir);
    }
    /* parent */
    close(fd);
    wrap_exitcode(child);

    unlink(fnmsg.s);
    _exit(0);
  }
}
